Kickstarter-iOS/Views/LoadingButton.swift

```diff
@@ -0,0 +1,159 @@
+import Library
+import Prelude
+import UIKit
+
+final class LoadingButton: UIButton {
+  // MARK: - Properties
+
+  private let viewModel: LoadingButtonViewModelType = LoadingButtonViewModel()
+
+  private lazy var activityIndicator: UIActivityIndicatorView = {
+    UIActivityIndicatorView(style: .white)
+      |> \.hidesWhenStopped .~ true
+      |> \.translatesAutoresizingMaskIntoConstraints .~ false
+  }()
+
+  public var isLoading: Bool = false {
+    didSet {
+      self.viewModel.inputs.isLoading(self.isLoading)
+    }
+  }
+
+  private var originalTitles: [UInt: String] = [:]
+
+  // MARK: - Lifecycle
+
+  override init(frame: CGRect) {
+    super.init(frame: frame)
+
+    self.configureViews()
+    self.bindViewModel()
+  }
+
+  required init?(coder: NSCoder) {
+    super.init(coder: coder)
+
+    self.configureViews()
+    self.bindViewModel()
+  }
+
+  override func setTitle(_ title: String?, for state: UIControl.State) {
+    // Do not allow changing the title while the activity indicator is animating
+    guard !self.activityIndicator.isAnimating else { return }
+
+    super.setTitle(title, for: state)
+  }
+
+  // MARK: - Configuration
+
+  private func configureViews() {
+    self.addSubview(self.activityIndicator)
+    self.activityIndicator.centerXAnchor.constraint(equalTo: self.centerXAnchor).isActive = true
+    self.activityIndicator.centerYAnchor.constraint(equalTo: self.centerYAnchor).isActive = true
+  }
+
+  // MARK: - View model
+
+  override func bindViewModel() {
+    super.bindViewModel()
+
+    self.viewModel.outputs.isUserInteractionEnabled
+      .observeForUI()
+      .observeValues { [weak self] isUserInteractionEnabled in
+        _ = self
+          ?|> \.isUserInteractionEnabled .~ isUserInteractionEnabled
+      }
+
+    self.viewModel.outputs.startLoading
+      .observeForUI()
+      .observeValues { [weak self] in
+        self?.startLoading()
+      }
+
+    self.viewModel.outputs.stopLoading
+      .observeForUI()
+      .observeValues { [weak self] in
+        self?.stopLoading()
+      }
+  }
+
+  // MARK: - Loading
+
+  private func startLoading() {
+    self.removeTitle()
+
+    self.activityIndicator.startAnimating()
+  }
+
+  private func stopLoading() {
+    self.activityIndicator.stopAnimating()
+
+    self.restoreTitle()
+  }
+
+  // MARK: - Titles
+
+  private func removeTitle() {
+    let disabledState = UIControl.State.disabled
+    let highlightedState = UIControl.State.highlighted
+    let normalState = UIControl.State.normal
+    let selectedState = UIControl.State.selected
+
+    if let disabledTitle = self.title(for: disabledState) {
```

Comment:
Extracting the title first will make sure that if for some reason we want to start the label with a loading state, we don't just clean the titles, but rather only cache previous titles if there were any.